### PR TITLE
Added rule 8740

### LINKED ISSUE #91: Rule 8740 : For a Section 47 Enquiry that has not held the Initial Child Protection Conference by the end of the census year, the start date must fall within the census year
If a \<Section47\> group does not contain the \<DateOfInitialCPC\> (N00110) and \<ICPCnotRequired\> (N00111) is false then the \<S47ActualStartDate\> (N00148) must be on or between [Start_Of_Census_Year] and \<ReferenceDate\> (N00603)

### PR DESCRIPTION
**Pull request by SLornieCYC**
Closes #91

### CODE REVIEW

**Comment by SLornieCYC** [comment]
Hi @tab1tha can you help me with what I should do here.  

This is currently failing because I have multiple S47s in-error for the same `LAchildID` with the same values for `S47ActualStartDate` and `ICPCnotRequired` (which can and does happen in real-world data!), which I expected to return as two rows in `assert len(issue_rows)` and `expected_df`.  

However the groupby makes this return as one row with `ROW_ID` = [4, 5].  I'm not sure whether that's allowed?  Can I just change my test output to match what is being returned by the function?

image

There are no other fields I can add to the zipped `link_id` to increase uniqueness of the rows here. `DateOfInitialCPC` won't help because for in-error records this will always be null, and there's no point adding `CINdetailsID` either because both S47s could be in the same CIN episode.

**Comment by willLPD2I** [comment]
Hi @SLornieCYC I've spent a while tinkering with this. Even though in real data someone shouldn't have two S47 start dates on the same date, because your sample data did, the df_issues groupby statement thought they were the same as they had the same link id, as link id is made of LAchildID and s47startdate in your rule. 

There are two fixes to this, I think, one is to add CINdetailsID to the groupby statement so the rule knows that, as the CINdetailsIDs will be different, they're not from the same S47, OR, add CINdetailsID to the link id. The first option works (i've tried) but isn't what you should do because it means issues_df isn't in the format the rest fot he backend wants, so, putting CINdetailsID as a third part of the tuple that make sup link id is best. It works in the code, I've tried it, but there might be something to do with CS data that means this isn't a viable solution?  I'd love to hear your thoughts.

**Comment by willLPD2I** [comment]
As an addendum to that comment, my solution doesn't work if you can have an S47 without a CINdetailsID, or if you can have the same CINdetailsID with different s47s?

**Comment by SLornieCYC** [comment]
Hi @willLPD2I although technically I agree that a child would not normally have multiple S47s on the same day that doesn't mean that combination can't happen in our raw data, which this tool needs to accommodate.

For instance in our CMS if an S47 has two ICPC next actions for the same child (which shouldn't happen but occassionally does - DQ!) the S47 would be returned twice in our CIN S47 data (one row per ICPC).  If neither ICPC had happened at that point the S47 could return in the way I have included in the test (e.g. if the S47 was dated towards the end of the previous year and neither of the two incoming ICPCs had been completed yet).  I admit it's quite an edge case!

In terms of your suggested use of CINdetailsID and your addendum:  
- You can definitely have more than one S47 within the same CINdetailsID (and if they have the same start date I would actually expect that to be the case, at least in our data).
- Depending on how ones reporting is configured to map the referral for CINdetailsID (and the underlying data quality!) it is equally possible to end up with S47s that don't have a CINdetailsID - ditto for content in other module in fact.
- Again depending on ones reporting configuration (this case doesn't apply to us) it might even be possible for the same S47 to be returned against multiple CINdetailsIDs if the referral episode data is wrong (e.g. overlapping episodes).

**Comment by tab1tha** [comment]
Hi @SLornieCYC, @willLPD2I 

Apologies. I was not able to respond sooner. 
This is okay. 

I'll add more comments in the file to explain why the result comes out as [4, 5] combined. There's a line that groups together all rows that have the same ERROR_ID.

File: cin_validator/rules/cin2022_23/rule_8740.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    # then the <S47ActualStartDate> (N00148) must be on or between [Start_Of_Census_Year] and <ReferenceDate> (N00603)
+    condition = (
+        df[DateOfInitialCPC].isna()
+        & (df[ICPCnotRequired].astype(str) == "0")
+        & (
+            (df[S47ActualStartDate] < collection_start)
+            | (df[S47ActualStartDate] > reference_date)
+        )
+    )
+    # get all the data that fits the failing condition. Reset the index so that ROW_ID now becomes a column of df
```
Comment (tab1tha) [inline cin_validator/rules/cin2022_23/rule_8740.py:62]: Just how it is supposed to be.
This is good work.

File: cin_validator/rules/cin2022_23/rule_8740.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    # - If we use the LAchildID-CPPstartDate combination, that artificially cancels out the instances where a start date repeats for the same child.
+    # Another rule checks for that condition. Not this one.
+    # - It is very unlikely that a combination of LAchildID-CPPstartDate-CPPendDate will repeat in the DataFrame.
+    # Hence, it can be used as a unique identifier of the row.
+
+    # Replace CPPstartDate and CPPendDate below with the columns concerned in your rule.
+    link_id = tuple(zip(df_issues[LAchildID], df_issues[S47ActualStartDate]))
+    df_issues["ERROR_ID"] = link_id
+    print("\n df_issues with link_id \n", df_issues)
+    df_issues = df_issues.groupby("ERROR_ID")["ROW_ID"].apply(list).reset_index()
```
Comment (tab1tha) [inline cin_validator/rules/cin2022_23/rule_8740.py:78]: Before this line (line 78), df_issues looks exactly like you have made it to be in your expected df.
Line 78 then groups together all the ROW_IDs that have the same ERROR_ID such that df_issues now becomes a dataframe of 

> {
                "ERROR_ID": (
                    "child1",
                    pd.to_datetime("26/05/1999", format="%d/%m/%Y", errors="coerce"),
                ),
                "ROW_ID": [4,5],
            },

Since the ERROR_IDs formed from both rows of your failing data are the same.

File: cin_validator/rules/cin2022_23/rule_8740.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+                "ROW_ID": [4],
+            },
+            {
+                "ERROR_ID": (
+                    "child1",
+                    pd.to_datetime("26/05/1999", format="%d/%m/%Y", errors="coerce"),
+                ),
+                "ROW_ID": [5],
+            },
+        ]
```
Comment (tab1tha) [inline cin_validator/rules/cin2022_23/rule_8740.py:192]: The ROW_IDs are grouped according to their ERROR_IDs so this will be:

> {
"ERROR_ID": (
"child1",
pd.to_datetime("26/05/1999", format="%d/%m/%Y", errors="coerce"),
),
"ROW_ID": [4,5],
},

**Review by tab1tha — CHANGES_REQUESTED** [review]
Just a few finishing touches left. 
We're almost there :)

File: cin_validator/rules/cin2022_23/rule_8740.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    expected_df = pd.DataFrame(
+        [
+            {
+                "ERROR_ID": (
+                    "child1",
+                    pd.to_datetime("26/05/1999", format="%d/%m/%Y", errors="coerce"),
+                ),
+                "ROW_ID": [4, 5],
+            },
+        ]
```
Comment (tab1tha) [inline cin_validator/rules/cin2022_23/rule_8740.py:182]: It's okay now. 

In the same way, if the child whose data failed on index 4 was different from the child whose data failed on index 5, then the groupby will not group them together and the two separate dataframe rows will be generated as shown:
{
"ERROR_ID": (
"child1",
pd.to_datetime("26/05/1999", format="%d/%m/%Y", errors="coerce"),
),
"ROW_ID": [4,5],
},
{
"ERROR_ID": (
"child2",
pd.to_datetime("26/05/1999", format="%d/%m/%Y", errors="coerce"),
),
"ROW_ID": [5],
},

**Review by tab1tha — APPROVED** [review]
Good to go.
Thank you @SLornieCYC

**Comment by tab1tha** [comment]
@amyrichards0n I've added you as a second reviewer. 
We need two approving reviews in order to merge. Could you please give this a look?

**Comment by tab1tha** [comment]
As we said on slack, we'll merge after 24 hours even if a second review is pending.
